### PR TITLE
Use case MLflow (the possums)

### PR DESCRIPTION
**Pull request by TJaniF**

### CODE REVIEW

File: learn/use-case-airflow-mlflow.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+![Graph view of the train DAG showing tasks fetching data and experiment information and training the model. Afterwards a task group contains tasks to register the model with MLflow, create a model version and transition the model version.](/img/examples/use-case-airflow-mlflow_train_dag_graph.png)
+
+The [`predict`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/predict.py) DAG uses the trained model from the previous DAG to create predictions and plots them against the target values. 
+
+![Graph view of the predict DAG showing the model and features being fetched. Afterwards the ModelLoadAndPredictOperator is used to run predictions with are plotted and saved as a file.](/img/examples/use-case-airflow-mlflow_predict_dag_graph.png)
+
+Note that the model is trained on the whole data and predictions are made on the same data, in a real world scenario with more data you'd want to split the data into a training, validation and test set.
```
Comment (kentdanas) [inline learn/use-case-airflow-mlflow.md:73]:
```suggestion
Note that the model is trained on the whole dataset and predictions are made on the same data. In a real world scenario you'd want to split the data into a training, validation and test set.
```

File: learn/use-case-airflow-mlflow.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    hyper_parameters={"alphas": np.logspace(-3, 1, num=30)},
+    run_name="RidgeCV",
+)
+```
+
+The run of the RidgeCV model can be viewed in the MLflow UI at `localhost:5000`.
+
+![Experiments tab of the MLflow UI showing the Possum_tails experiment with the RidgeCV run.](/img/examples/use-case-airflow-mlflow_experiments_ridgecv.png)
+
+Lastly the model training DAG registers the model and its version with MLflow using three operators from the MLflow Airflow provider. Note how information like the `run_id` or `version` of the model is pulled from XCom using [Jinja templates](templating.md).
```
Comment (kentdanas) [inline learn/use-case-airflow-mlflow.md:217]:
```suggestion
Lastly, the model training DAG registers the model and its version with MLflow using three operators from the MLflow Airflow provider. Note how information like the `run_id` or `version` of the model is pulled from XCom using [Jinja templates](templating.md).
```

**Review by kentdanas — APPROVED** [review]
@TJaniF This looks great! I don't really have many comments, I think good to copy edit

File: learn/use-case-airflow-mlflow.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+![Graph view of the train DAG showing tasks fetching data and experiment information and training the model. Afterwards a task group contains tasks to register the model with MLflow, create a model version and transition the model version.](/img/examples/use-case-airflow-mlflow_train_dag_graph.png)
+
+The [`predict`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/predict.py) DAG uses the trained model from the previous DAG to create predictions and plots them against the target values. 
+
+![Graph view of the predict DAG showing the model and features being fetched. Afterwards the ModelLoadAndPredictOperator is used to run predictions with are plotted and saved as a file.](/img/examples/use-case-airflow-mlflow_predict_dag_graph.png)
+
+Note that the model is trained on the whole dataset and predictions are made on the same data. In a real world scenario you'd want to split the data into a training, validation and test set.
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:73]:
```suggestion
Note that the model is trained on the whole dataset and predictions are made on the same data. In a real world scenario you'd want to split the data into a training, validation, and test set.
```

File: learn/use-case-airflow-mlflow.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+### Project code
+
+This use case shows many Airflow features and ways to interact with MLflow. The following sections will highlight a couple of relevant code snippets in each DAG and explain them in more detail.
+
+#### Feature engineering DAG
+
+The feature engineering DAG starts with a task that creates the necessary object storage buckets in the resource provided as `AWS_CONN_ID` using the [S3CreateBucketOperator](https://registry.astronomer.io/providers/apache-airflow-providers-amazon/versions/latest/modules/S3CreateBucketOperator). By default the project uses a local MinIO instance which is created when starting the Astro project. If you want to use remote object storage, you can change the `AWS_CONN_ID` in the `.env` file and provide your AWS credentials credentials.
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:81]:
```suggestion
The feature engineering DAG starts with a task that creates the necessary object storage buckets in the resource provided as `AWS_CONN_ID` using the [S3CreateBucketOperator](https://registry.astronomer.io/providers/apache-airflow-providers-amazon/versions/latest/modules/S3CreateBucketOperator). By default, the project uses a local MinIO instance, which is created when starting the Astro project. If you want to use remote object storage, you can change the `AWS_CONN_ID` in the `.env` file and provide your AWS credentials credentials.
```

File: learn/use-case-airflow-mlflow.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+The operator is [dynamically mapped](dynamic-tasks.md) over a list of bucket names to create all buckets in parallel.
+
+```python
+create_buckets_if_not_exists = S3CreateBucketOperator.partial(
+    task_id="create_buckets_if_not_exists",
+    aws_conn_id=AWS_CONN_ID,
+).expand(bucket_name=[DATA_BUCKET_NAME, MLFLOW_ARTIFACT_BUCKET, XCOM_BUCKET])
+```
+
+The `prepare_mlflow_experiment` [task group](task-groups.md) contains a pattern that lists all existing experiments in the MLflow instance connected via the `MLFLOW_CONN_ID` and creates a new experiment with a specified name if it does not exist yet using the [@task.branch decorator](airflow-branch-operator.md). The [MLflowClientHook](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/hooks/client.py) contains the `run` method that creates the new experiment by making a call to the MLflow API.
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:91]:
```suggestion
The `prepare_mlflow_experiment` [task group](task-groups.md) contains a pattern that lists all existing experiments in the MLflow instance connected via the `MLFLOW_CONN_ID`. It also creates a new experiment with a specified name if it does not exist yet using the [@task.branch decorator](airflow-branch-operator.md). The [MLflowClientHook](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/hooks/client.py) contains the `run` method that creates the new experiment by making a call to the MLflow API.
```

File: learn/use-case-airflow-mlflow.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+    feature_df = context["ti"].xcom_pull(
+        dag_id="feature_eng", task_ids="build_features", include_prior_dates=True
+    )
+    return feature_df
+```
+
+The ID number of the MLflow experiment is retrieved using the MLflowClientHook in the `fetch_experiment_id` task in order to track model training in the same experiment. 
+
+The `train_model` task, defined with the `@aql.dataframe` decorator shows how model training can be parameterized when using Airflow. In this example the hyperparameters, the `target_colum` and the model class are hardcoded but they could also be retrieved from upstream tasks via [XCom](airflow-passing-data-between-tasks.md) or passed into manual runs of the DAG via [DAG params](airflow-params.md).
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:176]:
```suggestion
The `train_model` task, defined with the `@aql.dataframe` decorator, shows how model training can be parameterized when using Airflow. In this example, the hyperparameters, the `target_colum`, and the model class are hardcoded, but they could also be retrieved from upstream tasks via [XCom](airflow-passing-data-between-tasks.md) or passed into manual runs of the DAG via [DAG params](airflow-params.md).
```

File: learn/use-case-airflow-mlflow.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    name=REGISTERED_MODEL_NAME,
+    version="{{ ti.xcom_pull(task_ids='register_model.create_model_version')['model_version']['version'] }}",
+    stage="Staging",
+    archive_existing_versions=True,
+)
+```
+
+You can view the registered models in the **Models** tab of the MLflow UI at `localhost:5000`.
+
+![Models tab of the MLflow UI showing the registered RidgeCV model having 3 Versions with the latest being in the Staging Stage.](/img/examples/use-case-airflow-mlflow_models_ridgecv.png)
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:251]:
```suggestion
![Models tab of the MLflow UI showing that the registered RidgeCV model has 3 Versions with the latest being in the Staging Stage.](/img/examples/use-case-airflow-mlflow_models_ridgecv.png)
```

File: learn/use-case-airflow-mlflow.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+#### Prediction DAG
+
+After retrieving the feature dataframe, the target column and the `model_run_id` from XCom, the `run_prediction` task uses the [ModelLoadAndPredictOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/pyfunc.py) to run a prediction on the whole dataset using the latest version of the registered RidgeCV model.
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:255]:
```suggestion
After retrieving the feature dataframe, the target column, and the `model_run_id` from XCom, the `run_prediction` task uses the [ModelLoadAndPredictOperator](https://github.com/astronomer/airflow-provider-mlflow/blob/main/mlflow_provider/operators/pyfunc.py) to run a prediction on the whole dataset using the latest version of the registered RidgeCV model. 
```

**Review by lzdanski — COMMENTED** [review]
What a fun example! I think my feedback is mostly about comma use - there's a couple consistent comma things so a quick style/grammar note:

- Technical writing/docs conventionally use the Oxford Comma. There's enough documented cases of ambiguity causing problems for organizations who didn't use the Oxford Comma, that it's become a pretty standard practice. 
- **That vs which** - Grammatically speaking, they technically have different functions, but in practice they're used interchangeably. However, "which" indicates a clause that's "optional" and appears at the beginning of a phrase that functions as a modifier/adjective/adverb, so it gets a comma before the "which". You can think of "which" as a flag that what follows is bonus information.

**Review by TJaniF — COMMENTED** [review]
Thank you for the review @lzdanski !

And for the grammar notes! Commas are definitely an area where I uhm... have room for improvement 😅 . I did not realize that difference between that and which, thanks!

File: learn/use-case-airflow-mlflow.md [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+title: "Predict possum tail length using linear regression with MLflow and Airflow"
+description: "Use Airflow and MLflow to conduct and track a regression model."
+id: use-case-airflow-mlflow
+sidebar_label: "Regression with Airflow + MLflow"
+sidebar_custom_props: { icon: 'img/integrations/mlflow.png' }
+---
+
+[MLflow](https://mlflow.org/) is a popular tool for tracking and managing machine learning models. It can be used together with Airflow for ML orchestration (MLOx), leveraging both tools for what they do best.
+
+This use case shows how to use MLflow with Airflow to engineer machine learning features, train a [scikit-learn Ridge linear regression model](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html) and create predictions based on the trained model.
```
Comment (lzdanski) [inline learn/use-case-airflow-mlflow.md:11]:
```suggestion
This use case shows how to use MLflow with Airflow to engineer machine learning features, train a [scikit-learn Ridge linear regression model](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html), and create predictions based on the trained model.
```

**Review by lzdanski — APPROVED** [review]
One last sneaky comma! But changes LGTM in preview.

File: learn/use-case-airflow-mlflow.md [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+### Data source
+
+This example uses the [Possum Regression dataset](https://www.kaggle.com/datasets/abrambeyer/openintro-possum) from Kaggle. It contains measurements of different attributes of 104 possums, such as total length, skull width, or age. This data was originally published by Lindenmayer et al. (1995) in the Australian Journal of Zoology and is commonly used to teach linear regression.
+
+### Project overview
+
+This project consists of three DAGs, which depend on each other using [Airflow datasets](airflow-datasets.md).
+
+![Datasets view of the use case project showing the DAG feature_eng updating the datasets astro+s://data/possum.csv and s3://data_possum.csv. The train DAG is scheduled to run as soon as the s3://data_possum.csv is updated and updates the model_trained dataset. The predict DAG is scheduled to run on updates to the model_trained dataset and updates the astro+s3://data/possum_tail_length.csv](/img/examples/use-case-airflow-mlflow_datasets_view.png)
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:59]: Putting us all to shame with these alt texts, I need to step up my game!

File: learn/use-case-airflow-mlflow.md [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+The [`feature_eng`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/feature_eng.py) DAG prepares the MLflow experiment and builds prediction features from the possum data.
+
+![Graph view of the feature_eng DAG showing a task creating the necessary MinIO buckets, a task group that prepares the MLflow experiment if it does not yet exist and data extraction and feature building tasks.](/img/examples/use-case-airflow-mlflow_feature_eng_dag_graph.png)
+
+The [`train`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/train.py) DAG trains a RidgeCV model on the features engineered in the previous DAG and then registers the model with MLflow using operators from the [MLflow Airflow provider](https://github.com/astronomer/airflow-provider-mlflow).
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:65]:
```suggestion
The [`train`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/train.py) DAG trains a RidgeCV model on the engineered features from `feature_eng` and then registers the model with MLflow using operators from the [MLflow Airflow provider](https://github.com/astronomer/airflow-provider-mlflow).
```

File: learn/use-case-airflow-mlflow.md [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+
+![Graph view of the train DAG showing tasks fetching data and experiment information and training the model. Afterwards a task group contains tasks to register the model with MLflow, create a model version and transition the model version.](/img/examples/use-case-airflow-mlflow_train_dag_graph.png)
+
+The [`predict`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/predict.py) DAG uses the trained model from the previous DAG to create predictions and plots them against the target values.
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:69]:
```suggestion
The [`predict`](https://github.com/astronomer/use-case-mlflow/blob/main/dags/predict.py) DAG uses the trained model from `train` to create predictions and plot them against the target values. 
```

File: learn/use-case-airflow-mlflow.md [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+            "artifact_location": f"s3://{artifact_bucket}/",
+        },
+    ).json()
+
+    return new_experiment_information
+```
+
+Feature engineering in the `build_features` task uses [Pandas for one hot encoding](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html) of categorical features and [scikit-learn to scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) numeric features.
+The [mlflow](https://pypi.org/project/mlflow/) package is used to [track](https://mlflow.org/docs/latest/python_api/mlflow.html?highlight=start_run#mlflow.start_run) the scaler run in MLflow.
+The task is defined using the [`@aql.dataframe`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/dataframe.html) decorator from the [Astro Python SDK](astro-python-sdk.md).
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:113]:
```suggestion

The task is defined using the [`@aql.dataframe`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/dataframe.html) decorator from the [Astro Python SDK](astro-python-sdk.md).
```

File: learn/use-case-airflow-mlflow.md [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        request_params={
+            "name": experiment_name,
+            "artifact_location": f"s3://{artifact_bucket}/",
+        },
+    ).json()
+
+    return new_experiment_information
+```
+
+Feature engineering in the `build_features` task uses [Pandas for one hot encoding](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html) of categorical features and [scikit-learn to scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) numeric features.
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:111]:
```suggestion
The `build_features` task completes feature engineering using [Pandas](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html) to one-hot encode categorical features and [scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) to scale numeric features.
```

File: learn/use-case-airflow-mlflow.md [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+The [mlflow](https://pypi.org/project/mlflow/) package is used to [track](https://mlflow.org/docs/latest/python_api/mlflow.html?highlight=start_run#mlflow.start_run) the scaler run in MLflow.
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:112]:
```suggestion

The [mlflow](https://pypi.org/project/mlflow/) package is used to [track](https://mlflow.org/docs/latest/python_api/mlflow.html?highlight=start_run#mlflow.start_run) the scaler run in MLflow.
```

File: learn/use-case-airflow-mlflow.md [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+
+```
+
+You can view the Scaler run in the MLflow UI at `localhost:5000`.
+
+![Experiments tab of the MLflow UI showing the Possum_tails experiment with the Scaler run.](/img/examples/use-case-airflow-mlflow_experiments_scaler.png)
+
+#### Model training DAG
+
+The model training DAG is scheduled to run as soon as the last task in the feature engineering DAG completes using a [dataset](airflow-datasets.md).
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:150]:
```suggestion
Airflow [datasets](airflow-datasets.md) let you schedule DAGs based on when a specific file or database is updated in a separate DAG. In this example, the model training DAG is scheduled to run as soon as the last task in the feature engineering DAG completes.
```

File: learn/use-case-airflow-mlflow.md [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+
+    feature_df = context["ti"].xcom_pull(
+        dag_id="feature_eng", task_ids="build_features", include_prior_dates=True
+    )
+    return feature_df
+```
+
+The ID number of the MLflow experiment is retrieved using the MLflowClientHook in the `fetch_experiment_id` task in order to track model training in the same experiment. 
+
+The `train_model` task, defined with the `@aql.dataframe` decorator, shows how model training can be parameterized when using Airflow. In this example, the hyperparameters, the `target_colum`, and the model class are hardcoded, but they could also be retrieved from upstream tasks via [XCom](airflow-passing-data-between-tasks.md) or passed into manual runs of the DAG via [DAG params](airflow-params.md).
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:176]:
```suggestion
The `train_model` task, defined with the `@aql.dataframe` decorator, shows how model training can be parameterized when using Airflow. In this example, the hyperparameters, the `target_colum`, and the model class are hardcoded, but they could also be retrieved from upstream tasks via [XCom](airflow-passing-data-between-tasks.md) or passed into manual runs of the DAG using [DAG params](airflow-params.md).
```

File: learn/use-case-airflow-mlflow.md [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+    feature_df=fetched_feature_df,
+    experiment_id=fetched_experiment_id,
+    target_column=TARGET_COLUMN,
+    model_class=RidgeCV,
+    hyper_parameters={"alphas": np.logspace(-3, 1, num=30)},
+    run_name="RidgeCV",
+)
+```
+
+The run of the RidgeCV model can be viewed in the MLflow UI at `localhost:5000`.
```
Comment (jwitz) [inline learn/use-case-airflow-mlflow.md:213]:
```suggestion
You can view the run of the RidgeCV model in the MLflow UI at `localhost:5000`.
```

**Review by jwitz — APPROVED** [review]
Finally!! 

image

This example includes so many cool features in one package! Not the main focus here, but I had no idea you could use an AWS connection to connect to a locally running container

File: learn/use-case-airflow-mlflow.md (hunk 10 above)
Comment (TJaniF) [reply learn/use-case-airflow-mlflow.md:61]: 😊 thank you (also I will have to resist not to put jokes and sneaky remarks in all of my alt texts, so tempting)

**Review by TJaniF — COMMENTED** [review]
😍 baby possums! 

Thank you @jwitz ! And yes I did not realize that either for way too long, I once wrote a couple of operators for MinIO, not realizing I could just plug in the S3 ones 😅

**Pull request merged**